### PR TITLE
RavenDB-20180: fix NRE after restoring subscription from dump file

### PR DESCRIPTION
**Pull request by garayx**
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-20180

### Additional description

On importing subscription from dump file to sharded database, we were creating it using the regular `PutSubscriptionCommand` which doesn't add the correct properties for sharding database, therefore we got NRE afterwards.

### Type of change

- Bug fix

### How risky is the change?

- Low 

### Backward compatibility

- Non breaking change

### Is it platform specific issue?

- No

### Documentation update

- No documentation update is needed 

### Testing by Contributor

- Tests have been added that prove the fix is effective or that the feature works
- It has been verified by manual testing

### Testing by RavenDB QA team

- No special testing by RavenDB QA team is needed

### Is there any existing behavior change of other features due to this change?

- No

### UI work

- No UI work is needed

### CODE REVIEW

File: src/Raven.Server/ServerWide/Commands/Sharding/PutShardedSubscriptionCommand.cs [hunk 1]
```diff
@@ -45,6 +46,14 @@ protected override void AssertValidChangeVector()
         // the CVs were validated in handler
     }
 
+    protected override void AssertCanPutSubscription(RawDatabaseRecord record)
+    {
+        if (record.IsSharded == false)
+        {
+            throw new PutSubscriptionException($"'{nameof(PutShardedSubscriptionCommand)}' is not supported for regular database.");
```
Comment (karmeli87) [inline src/Raven.Server/ServerWide/Commands/Sharding/PutShardedSubscriptionCommand.cs]: Please make sure that this kind of exception will not cause the state machine to be stuck.

Comment (karmeli87) [reply src/Raven.Server/ServerWide/Commands/Sharding/PutShardedSubscriptionCommand.cs]: It should be part of the excepted exception (or be wrapped in such)

Comment (garayx) [reply src/Raven.Server/ServerWide/Commands/Sharding/PutShardedSubscriptionCommand.cs]: https://github.com/ravendb/ravendb/blob/90042221c1e80b58722fcfa8d76f2ee2659ad70c/src/Raven.Server/ServerWide/ClusterStateMachine.cs#L882

File: src/Raven.Client/Exceptions/Documents/Subscriptions/PutSubscriptionException.cs [hunk 2]
```diff
@@ -0,0 +1,15 @@
+using System;
+
+namespace Raven.Client.Exceptions.Documents.Subscriptions
+{
+    public class PutSubscriptionException : SubscriptionException
```
Comment (ppekrol) [inline src/Raven.Client/Exceptions/Documents/Subscriptions/PutSubscriptionException.cs]: SubscriptionCreationException -> let's be consistent with other exceptions

File: src/Raven.Server/ServerWide/Commands/Subscriptions/PutSubscriptionBatchCommand.cs [hunk 3]
```diff
@@ -1,18 +1,20 @@
...
 {
-    public class PutSubscriptionBatchCommand : CommandBase
+    public abstract class PutSubscriptionBatchCommandBase<T> : CommandBase
+    where T : PutSubscriptionCommand
     {
-        public List<PutSubscriptionCommand> Commands;
+        public List<T> Commands;
 
-        public PutSubscriptionBatchCommand()
+        public PutSubscriptionBatchCommandBase()
```
Comment (ppekrol) [inline src/Raven.Server/ServerWide/Commands/Subscriptions/PutSubscriptionBatchCommand.cs]: protected

File: src/Raven.Server/ServerWide/Commands/Subscriptions/PutSubscriptionBatchCommand.cs [hunk 4]
```diff
@@ -30,4 +32,28 @@ public override DynamicJsonValue ToJson(JsonOperationContext context)
             return djv;
         }
     }
+
+    public class PutSubscriptionBatchCommand : PutSubscriptionBatchCommandBase<PutSubscriptionCommand>
```
Comment (ppekrol) [inline src/Raven.Server/ServerWide/Commands/Subscriptions/PutSubscriptionBatchCommand.cs:5]: move to a different file

File: src/Raven.Server/ServerWide/Commands/Subscriptions/PutSubscriptionBatchCommand.cs [hunk 5]
```diff
@@ -30,4 +32,28 @@ public override DynamicJsonValue ToJson(JsonOperationContext context)
...
+        public PutSubscriptionBatchCommand()
+        {
+        }
+
+        public PutSubscriptionBatchCommand(List<PutSubscriptionCommand> commands, string uniqueRequestId) : base(commands, uniqueRequestId)
+        {
+        }
+    }
+
+    public class PutShardedSubscriptionBatchCommand : PutSubscriptionBatchCommandBase<PutShardedSubscriptionCommand>
```
Comment (ppekrol) [inline src/Raven.Server/ServerWide/Commands/Subscriptions/PutSubscriptionBatchCommand.cs]: this is not a place for that, what is the benefit of having this command?

File: src/Raven.Server/Smuggler/Documents/Actions/DatabaseSubscriptionActions.cs [hunk 6]
```diff
@@ -2,57 +2,92 @@
...
     {
         if (_subscriptionCommands.Count == 0)
             return;
 
         await SendCommandsAsync();
     }
+}
 
-    public async ValueTask WriteSubscriptionAsync(SubscriptionState subscriptionState)
+public class ShardedDatabaseSubscriptionActions : DatabaseSubscriptionActionsBase<PutShardedSubscriptionCommand>
```
Comment (ppekrol) [inline src/Raven.Server/Smuggler/Documents/Actions/DatabaseSubscriptionActions.cs]: should be moved to appropriate place?

File: src/Raven.Server/ServerWide/Commands/Subscriptions/PutSubscriptionBatchCommand.cs (hunk 5 above)
Comment (garayx) [reply src/Raven.Server/ServerWide/Commands/Subscriptions/PutSubscriptionBatchCommand.cs]: we need a way to recognize if we add a batch of sharded subscriptions or  a batch of regular in CSM

**Pull request merged**